ls1intum/Artemis
# `Programming exercises`: Add Gradle as project type for Java exercises on Atlassian setup

### Checklist
This PR has quite a lot of line additions but this is mainly due to the newly added template files (like .gitignore, wrapper files, etc.). There are around 250 added lines of Java-/TypeScript-Code.
#### General
#### Server
#### Client
#### Changes affecting Programming Exercises
-> The new project type is only available on the Atlassian Setup. However, these changes should not cause any problems on the Jenkins/Gitlab-Setup. In addition, these project types should not be selectable on this setup.

### Motivation and Context
As part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA) in Java, testwise coverage has to be recorded. We will use a plugin that is only available for Gradle projects, therefore we decided to add Gradle as a new project type.
In addition, Maven is getting less innovative and may be considered as not state-of-the-art anymore.

### Description
This PR introduces two new project types:
- Plain - Gradle: The template and solution repositories are plain Java code, the test repository is a Gradle project. The Maven equivalent in Artemis is 'Eclipse' for which the template and solution repositories are also plain Java code but the test repository is a Maven project.
- Gradle - Gradle: The template, solution, and test repository are all Gradle projects. The Maven equivalent is 'Maven' where all three repositories are Maven projects

The creation of the project files has been adapted as well as the creation of the build plans on the Bamboo server.

### Related changes
In order to work, the following related changes have been made:
- Changing the Ares testing framework to support Gradle: [#191](https://github.com/ls1intum/Ares/pull/191)
- Cache the Gradle dependency in the artemis-maven-docker which is used for executing the builds on Bamboo/Jenkins: [#63](https://github.com/ls1intum/artemis-maven-docker/pull/63)

### Steps for Testing
Prerequisites:
- Testserver with the Atlassian Setup (TS1, TS3)
- 1 Instructor
- 1 Student

1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to Exercises
4. Generate new Java programming exercises. Check for both projects type 'Plain - Gradle' and 'Gradle - Gradle' that the following things are working correctly:
- creating the build plans and repositories
- Template Score: 0%, Solution Score: 100%
- importing as a Gradle project into the local IDE works fine
- when activating the static code analysis, introducing some style issues, and pushing the changes, the build result contains feedback on these style issues
- when applying and pushing changes to (one of the three) repositories, the test case results update accordingly and these results are as expected
- when pushing some code that does not compile, the build result shows a build error
- adding new dependencies to the build.gradle in the test repository and using classes from this dependency works fine
- when activating sequential test runs, confirm that the behavioral tests are only executed after the structural tests have all passed. Otherwise, no test results are shown for these behavioral tests.
- clone all three repositories locally. Import the test repository. Add the template and solution repositories as an existing module. Navigate in IntelliJ to "File" -> "Project Structure" -> "Modules" -> "$ProjectName-Tests" -> Sub-Option "test" -> "Dependencies" -> Button "+" -> "Module Dependency" -> Select the template or solution repository with appended ".main". Verify now that the Test classes "recognize" the code from the respective (template or solution) repository. Execute "./gradlew clean test" and check if all tests run as expected.
 
Only for project type 'Gradle - Gradle':
- Check that also the solution and template repository are Gradle projects
- Add a dependency to all three repositories and use classes from this dependency. When pushing the changes the result should be as expected.


5. Code editor: Build errors and style issues are displayed with a symbol in the corresponding line in the respective file.


### Review Progress

#### Code Review
#### Manual Tests

### Screenshots
![Bildschirmfoto 2022-03-10 um 20 04 39](https://user-images.githubusercontent.com/73833780/157736393-b237cb5d-178e-473e-a708-772734c25d9b.png)
![Bildschirmfoto 2022-03-10 um 20 04 49](https://user-images.githubusercontent.com/73833780/157736431-d9fbdb05-89cc-4eed-81eb-607c935310c8.png)

| File | Branch | Line |
|------------|-------:|-----:|
| ProgrammingExerciseService.java | 81% | 90.04% |
| FeedbackRepository.java | 81% | 86.9% |
| FileService.java | 86% | 89.7% |
